presto-druid/src/main/java/io/prestosql/plugin/druid/DruidJdbcClient.java

```diff
@@ -0,0 +1,365 @@
+import static io.prestosql.spi.type.VarcharType.createVarcharType;
+import static java.lang.String.format;
+import static java.sql.DatabaseMetaData.columnNoNulls;
+
+public class DruidJdbcClient
+        extends BaseJdbcClient
+{
+    // Druid maintains its datasources related metadata by setting the catalog name as "druid"
+    // Note that while a user may name the catalog name as something else, metadata queries made
+    // to druid will always have the TABLE_CATALOG set to DRUID_CATALOG
+    private static final String DRUID_CATALOG = "druid";
+    // All the datasources in Druid are created under schema "druid"
+    public static final String DRUID_SCHEMA = "druid";
+
+    @Inject
+    public DruidJdbcClient(BaseJdbcConfig config, ConnectionFactory connectionFactory)
+            throws SQLException
+    {
+        super(config, "\"", connectionFactory);
+    }
+
+    @Override
+    public List<SchemaTableName> getTableNames(JdbcIdentity identity, Optional<String> schema)
+    {
+        try (Connection connection = connectionFactory.openConnection(identity)) {
+            try (ResultSet resultSet = getTables(connection, schema, Optional.empty())) {
+                ImmutableList.Builder<SchemaTableName> list = ImmutableList.builder();
+                while (resultSet.next()) {
+                    list.add(new SchemaTableName(
+                            resultSet.getString("TABLE_SCHEM"),
+                            resultSet.getString("TABLE_NAME")));
+                }
+                return list.build();
+            }
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    protected Collection<String> listSchemas(Connection connection)
+    {
+        return ImmutableList.of(DRUID_SCHEMA);
+    }
+
+    /**
+     * Overriden to filter out tables that don't match @param schemaTableName
+     */
+    @Override
+    public Optional<JdbcTableHandle> getTableHandle(JdbcIdentity identity, SchemaTableName schemaTableName)
+    {
+        try (Connection connection = connectionFactory.openConnection(identity)) {
+            String jdbcSchemaName = schemaTableName.getSchemaName();
+            String jdbcTableName = schemaTableName.getTableName();
+            try (ResultSet resultSet = getTables(connection, Optional.of(jdbcSchemaName), Optional.of(jdbcTableName))) {
+                List<JdbcTableHandle> tableHandles = new ArrayList<>();
+                while (resultSet.next()) {
+                    tableHandles.add(new JdbcTableHandle(
+                            schemaTableName,
+                            DRUID_CATALOG,
+                            resultSet.getString("TABLE_SCHEM"),
+                            resultSet.getString("TABLE_NAME")));
+                }
+                if (tableHandles.isEmpty()) {
+                    return Optional.empty();
+                }
+                if (tableHandles.size() > 1) {
```

Comment:
the filtering should be applied regardless whether we have 1 candidate or many
(if we have 1, it may be a wrong one)